src/backend/d3d12/ShaderModuleD3D12.cpp

```diff
@@ -33,26 +33,24 @@ namespace backend { namespace d3d12 {
 
         ExtractSpirvInfo(compiler);
 
-        // rename bindings so that each register type b/u/t/s starts at 0 and then offset by
+        // rename bindings so that each register type c/u/t/s starts at 0 and then offset by
         // kMaxBindingsPerGroup * bindGroupIndex
-        auto RenumberBindings = [&](std::vector<spirv_cross::Resource> resources) {
-            std::array<uint32_t, kMaxBindGroups> baseRegisters = {};
-
-            for (const auto& resource : resources) {
-                auto bindGroupIndex =
-                    compiler.get_decoration(resource.id, spv::DecorationDescriptorSet);
-                auto& baseRegister = baseRegisters[bindGroupIndex];
-                auto bindGroupOffset = bindGroupIndex * kMaxBindingsPerGroup;
-                compiler.set_decoration(resource.id, spv::DecorationBinding,
-                                        bindGroupOffset + baseRegister++);
+        const auto& moduleBindingInfo = GetBindingInfo();
+        for (int group = 0; group < moduleBindingInfo.size(); ++group) {
+            const auto& groupBindingInfo = moduleBindingInfo[group];
+
+            static constexpr int kNumBindingTypes = 4;
+            std::array<uint32_t, kNumBindingTypes> baseRegisters{};
+
+            for (const auto& bindingInfo : groupBindingInfo) {
+                if (bindingInfo.used) {
+                    auto& baseRegister = baseRegisters[static_cast<uint32_t>(bindingInfo.type)];
```

Comment:
Can we have a switch instead of a static_cast, or at least have an assert that we are inside the array?

Also I think we should use auto a bit less: Chromium's style guide asks that simple types are spelled out.

Reply:
I wish we had an `enum_map` type. For now I added an assert.

Reply:
Actually changed it to a switch.